rvps/src/lib.rs

```diff
@@ -0,0 +1,126 @@
+// Copyright (c) 2022 Alibaba Cloud
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#![allow(clippy::new_without_default)]
+
+pub mod cache;
+pub mod extractors;
+pub mod pre_processor;
+pub mod reference_value;
+
+use std::time::SystemTime;
+
+use anyhow::{anyhow, Result};
+use chrono::{DateTime, Utc};
+use extractors::{Extractors, ExtractorsImpl};
+use pre_processor::{PreProcessor, PreProcessorAPI, Ware};
+use serde::{Deserialize, Serialize};
+
+pub use cache::Cache;
+pub use reference_value::{ReferenceValue, TrustedDigest};
+
+/// Default version of Message
+static MESSAGE_VERSION: &str = "0.1";
+
+/// Message is an overall packet that Reference Value Provider Service
+/// receives. It will contain payload (content of different provenance,
+/// JSON format), provenance type (indicates the type of the payload)
+/// and a version number (use to distinguish different version of
+/// message, for extendability).
+/// * `version`: version of this message.
+/// * `payload`: content of the provenance, JSON encoded.
+/// * `type`: provenance type of the payload.
+#[derive(Serialize, Deserialize, Debug)]
+pub struct Message {
+    #[serde(default = "default_version")]
+    version: String,
+    payload: String,
+    r#type: String,
+}
+
+/// Set the default version for Message
+fn default_version() -> String {
+    MESSAGE_VERSION.into()
+}
+
+/// The interfaces of Reference Value Provider Service
+/// * `verify_and_extract` is responsible for verify a message and
+/// store reference values from it.
+/// * `get_digests` gets trusted digests by the artifact's name.
+pub trait RVPSAPI {
+    fn verify_and_extract(&mut self, message: Message) -> Result<()>;
+    fn get_digests(&self, name: &str) -> Result<Option<TrustedDigest>>;
+}
+
+/// The core of the RVPS, s.t. componants except communication componants.
+pub struct Core<T: Cache> {
+    pre_processor: PreProcessor,
+    extractors: ExtractorsImpl,
+    cache: T,
+}
+
+impl<T: Cache> Core<T> {
+    /// Instantialize a new RVPS Core
+    pub fn new(cache: T) -> Self {
+        let pre_processor = PreProcessor::default();
+
+        let extractors = ExtractorsImpl::default();
+
+        Core {
+            pre_processor,
+            extractors,
+            cache,
+        }
+    }
+
+    /// Add Ware to the Core's Pre-Processor
+    pub fn with_ware(&mut self, ware: Box<dyn Ware>) -> &Self {
+        self.pre_processor.add_ware(ware);
+        self
+    }
+}
+
+impl<T: Cache> RVPSAPI for Core<T> {
+    fn verify_and_extract(&mut self, mut message: Message) -> Result<()> {
+        // Judge the version field
+        if message.version != MESSAGE_VERSION {
+            return Err(anyhow!(
```

Comment:
Could you add a unit test to assert this error occurs?

Reply:
Thanks to your suggestion, I think this error occurs when the writer edited the `version` field by mistake. If a different layout message is sent, certainly it can not be deserialized, thus not able to reach here.

I will add the unit test in the next pr. Because I didn't add any implementation for `Cache`, which is needed to create a RVPS's `Core` object, the test is now not able to be written.